File: src/main/java/com/nir/beans/Methods.java
package com.nir.beans;

import com.nir.utils.math.method.Method;
import com.nir.utils.math.method.generalized.GeneralizedMethodInfoJsonPojo;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

public class Methods {

    private static final GeneralizedMethodInfoComponent generalizedMethodInfoComponent = Beans.generalizedMethodInfoComponent();
    private static final GeneralizedMethodComponent generalizedMethodComponent = Beans.generalizedMethodComponent();
    private static final HardcodedMethodComponent hardcodedMethodComponent = Beans.hardcodedMethodComponent();
    private static final List<Method> generalizedMethods = new ArrayList<>();
    private static final List<Method> hardcodedMethods = new ArrayList<>();
    private static final List<Method> allMethods = new ArrayList<>();
    private static final List<GeneralizedMethodInfoJsonPojo> generalizedMethodInfoJsonPojos = new ArrayList<>();

    public static List<Method> getAll() {
        initGeneralizedMethods();
        initHardcodedMethods();
        initAllMethods();
        return allMethods;
    }

    public static List<String> getNames() {
        return getAll().stream().map(Method::getName).collect(Collectors.toList());
    }

    private static void initAllMethods() {
        if(allMethods.isEmpty()) {
            allMethods.addAll(hardcodedMethods);
            allMethods.addAll(generalizedMethods);
        }
    }

    public static Method getByName(String name) {
        final List<Method> all = getAll();
        final Optional<Method> first = all.stream().filter(it -> it.getName().equals(name)).findFirst();
        final Method method = first.get();
        return method;
    }

    private static void initHardcodedMethods() {
        if (hardcodedMethods.isEmpty()) {
            final List<Method> created = hardcodedMethodComponent.getAll();
            hardcodedMethods.addAll(created);
        }
    }

    private static void initGeneralizedMethods() {
        initGeneralizedMethodsInfos();
        if (generalizedMethods.isEmpty()) {
            final List<Method> created = generalizedMethodComponent.create(generalizedMethodInfoJsonPojos);
            generalizedMethods.addAll(created);
        }
    }

    private static void initGeneralizedMethodsInfos() {
        if (generalizedMethodInfoJsonPojos.isEmpty()) {
            final List<GeneralizedMethodInfoJsonPojo> all = generalizedMethodInfoComponent.getAll();
            generalizedMethodInfoJsonPojos.addAll(all);
        }
    }
}

File: src/main/kotlin/com/nir/beans/method/Methods.kt
package com.nir.beans.method

import com.nir.beans.Beans.methodConverterComponent
import com.nir.beans.Beans.methodInfoComponent
import com.nir.beans.Beans.hardcodedMethodComponent
import com.nir.beans.method.generalized.MethodInfoJsonPojo

object Methods {
    private val methodInfoComponent: MethodInfoComponent = methodInfoComponent()

    private val generalizedMethodComponent = methodConverterComponent()
    private val hardcodedMethodComponent = hardcodedMethodComponent()

    private val hardcodedMethods: MutableList<Method> = ArrayList()
    private val allMethods: MutableList<Method> = ArrayList()
    private val allMethodsJsonPojos: MutableList<MethodInfoJsonPojo> = ArrayList()

    init {

        if (hardcodedMethods.isEmpty()) {
            allMethods.addAll(hardcodedMethodComponent.getMethods)
        }

        allMethodsJsonPojos.addAll(methodInfoComponent.getMethods)
        allMethodsJsonPojos.forEach { println(it) }
    }

    @JvmStatic
    fun getByName(name: String): Method {
        val all = allMethods.filter { it.name == name }
        if (all.isNotEmpty()) {
            return all.first()
        }

        val pojo = allMethodsJsonPojos.first { it.name == name }
        val method = generalizedMethodComponent.create(arrayListOf(pojo))[0]
        if (!allMethods.contains(method)) {
            allMethods.add(method)
        }

        return method
    }
}